index.js

```diff
@@ -172,18 +53,154 @@ module.exports = function (fastify, opts, next) {
         cyclicDependencyCheck = {}
 
         try {
-          loadPlugin(allPlugins[pluginKeys[i]])
+          const plugin = allPlugins[pluginKeys[i]]
+          if (plugin == null) {
+            continue
+          }
+          loadPlugin(plugin)
         } catch (err) {
-          next(enrichError(err))
-          return
+          throw enrichError(err)
         }
       }
-
-      next()
     })
-  })
 }
 
 // do not create a new context, do not encapsulate
 // same as fastify-plugin
 module.exports[Symbol.for('skip-override')] = true
+
+function importPluginAndMetaData (file, opts, allPlugins, defaultPluginOptions) {
+  return importPlugin(file)
+    .then((plugin) => {
+      const pluginOptions = Object.assign({}, defaultPluginOptions)
+      const pluginMeta = plugin[Symbol.for('plugin-meta')] || {}
+      const pluginName = pluginMeta.name || file
+
+      if (opts && !plugin.autoPrefix) {
+        plugin.autoPrefix = opts.prefix
+      }
+
+      if (plugin.autoload === false) {
+        return null
+      }
+
+      if (plugin.autoPrefix) {
+        const prefix = pluginOptions.prefix || ''
+        pluginOptions.prefix = prefix + plugin.autoPrefix
+      }
+
+      if (plugin.prefixOverride !== undefined) {
+        pluginOptions.prefix = plugin.prefixOverride
+      }
+
+      if (allPlugins[pluginName]) {
+        throw new Error(`Duplicate plugin: ${pluginName}`)
+      }
+
+      allPlugins[pluginName] = {
+        plugin,
+        name: pluginName,
+        dependencies: pluginMeta.dependencies,
+        options: pluginOptions
+      }
+    })
+    .catch((err) => {
+      throw enrichError(err)
+    })
+}
+
+function importPlugin (file) {
+  if (file.endsWith('.mjs')) {
+    return import(file)
+  } else {
+    try {
+      return Promise.resolve(require(file))
+    } catch (err) {
+      return Promise.reject(err)
+    }
+  }
+}
+
+function loadPluginFiles (opts) {
+  const packagePattern = /^package\.json$/im
+  const indexPattern = opts.includeTypeScript
+    ? /^index\.(ts|js|mjs)$/im
+    : /^index\.(js|mjs)$/im
+  const scriptPattern = opts.includeTypeScript
+    ? /\.(ts|js|mjs)$/im
+    : /\.(js|mjs)$/im
+
+  return readDir(opts.dir)
+    .then((list) => {
+      const promises = []
+      for (const file of list) {
+        promises.push(loadPluginFile(opts, file, packagePattern, indexPattern, scriptPattern))
+      }
+
+      return Promise.all(promises).then(flat)
+    })
+}
+
+function loadPluginFile (opts, file, packagePattern, indexPattern, scriptPattern) {
+  if (opts.ignorePattern && file.match(opts.ignorePattern)) {
+    return Promise.resolve([{ skip: true }])
+  }
+  const toLoad = path.join(opts.dir, file)
+  return getStat(toLoad)
+    .then((stat) => {
```

Comment:
In general, I prefer to have these two in the same line `getStat(toLoad).then((stat) => {`. This will limit nesting and make call a bit easier to read.

Reply:
I changed that.